### ISSUE TITLE
Shutdown Scylla when writes to disk timeout

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#15642]
We saw in a user's cluster one of the nodes being in a strange situation - it was amassing a large backlog of queries, a large backlog of compactions that just grew and grew without ever going down, requests timed out - while the other nodes were normal (e.g., their compaction backlog was always 0 except for short durations). The kernel's log on that node indicated that something was badly wrong in the kernel or the hardware causing fsync to hang:

```
kern  :err   : [Thu Oct  5 02:07:50 2023] INFO: task kworker/0:3:1475956 blocked for more than 241 seconds.
kern  :err   : [Thu Oct  5 02:07:50 2023]       Not tainted 5.13.0-1031-aws #35~20.04.1-Ubuntu
kern  :err   : [Thu Oct  5 02:07:50 2023] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
kern  :info  : [Thu Oct  5 02:07:50 2023] task:kworker/0:3     state:D stack:    0 pid:1475956 ppid:     2 flags:0x00004000
kern  :info  : [Thu Oct  5 02:07:50 2023] Workqueue: events aio_fsync_work
kern  :info  : [Thu Oct  5 02:07:50 2023] Call Trace:
kern  :info  : [Thu Oct  5 02:07:50 2023]  <TASK>
kern  :info  : [Thu Oct  5 02:07:50 2023]  __schedule+0x2ee/0x900
kern  :info  : [Thu Oct  5 02:07:50 2023]  schedule+0x4f/0xc0
kern  :info  : [Thu Oct  5 02:07:50 2023]  __xfs_log_force_lsn+0x1a3/0x1e0 [xfs]
kern  :info  : [Thu Oct  5 02:07:50 2023]  ? wake_up_q+0xa0/0xa0
kern  :info  : [Thu Oct  5 02:07:50 2023]  xfs_log_force_lsn+0x94/0x100 [xfs]
kern  :info  : [Thu Oct  5 02:07:50 2023]  xfs_log_force_inode+0x6e/0x90 [xfs]
kern  :info  : [Thu Oct  5 02:07:50 2023]  xfs_dir_fsync+0x29/0x70 [xfs]
kern  :info  : [Thu Oct  5 02:07:50 2023]  vfs_fsync_range+0x49/0x80
kern  :info  : [Thu Oct  5 02:07:50 2023]  ? psi_task_switch+0x121/0x260
kern  :info  : [Thu Oct  5 02:07:50 2023]  vfs_fsync+0x1c/0x20
kern  :info  : [Thu Oct  5 02:07:50 2023]  aio_fsync_work+0x32/0x2e0
kern  :info  : [Thu Oct  5 02:07:50 2023]  process_one_work+0x220/0x3c0
kern  :info  : [Thu Oct  5 02:07:50 2023]  worker_thread+0x4d/0x3f0
kern  :info  : [Thu Oct  5 02:07:50 2023]  ? process_one_work+0x3c0/0x3c0
kern  :info  : [Thu Oct  5 02:07:50 2023]  kthread+0x12b/0x150
kern  :info  : [Thu Oct  5 02:07:50 2023]  ? set_kthread_struct+0x40/0x40
kern  :info  : [Thu Oct  5 02:07:50 2023]  ret_from_fork+0x22/0x30
kern  :info  : [Thu Oct  5 02:07:50 2023]  </TASK>
```

I don't know why this kernel error happened (we suspect a genuine hardware problem - rebooting the node caused its disk to completely disappear and supposedly this only happens on AWS if they detect faulty hardware), but that's not the point of this issue.
The point of this issue is that leaving the node in this broken state **for hours** devastated the performance of the cluster for hours. Scylla on this node was better-off to detect that it can't make any forward progress and automatically shut down. When we finally did shut down this node manually, the cluster finally went back to normal (just with one less node, that later had to be replaced). If this shutdown was automatic, the user would have avoided several hours of terrible cluster performance.

Scylla already has code to exit immediately when detecting a hard "IO error". It appears that in this case, we didn't get an actual IO error but all IO hung in some way. It would be good to be able to detect this case as well, and exit in this case as well.

Of course we need to be careful not to exit Scylla because of some temporary overload or something, so we need to find a way to detect permanent serious problems, and not exit on every transient delay. Perhaps we can keep a record of the last successful write and the last write or last failed write, and exit if writes have been failing for a significant time (e.g., 1 minute).

CC @avikivity

**Comment by avikivity** [comment]
Unfortunately this is a very difficult problem to tackle, and even more to test.

I/O can appear to hang due to overload, or due to a temporary network problem (for network attached storage), and I can probably think of more reasons.

Even worse is when you have a large slowdown. Then the detector thinks everything is okay but in reality the entire cluster is slowed down to the speed of the slowest disk or node.

I don't think we'll be able to make progress here.

**Comment by nyh** [comment]
> Unfortunately this is a very difficult problem to tackle, and even more to test.
> 
> I/O can appear to hang due to overload, or due to a temporary network problem (for network attached storage), and I can probably think of more reasons.

I agree this will be hard to test - e.g., we aren't really sure what went wrong on the node that caused the current problem so we don't know how to "fake" this problem in a test. However, I do think that we can fairly easily detect a situation where our low-level IO layer tried to send N operations in the last 5 minute (for example), but zero operations completed in the last 5 minute. 

If our node relies on network attached storage that hasn't worked for 5 minutes, we're pretty screwed anyway - but I agree that if this is felt across the entire cluster, it's better to lose availability for 5 minutes and then immediately restore operation, instead of shutting down all the nodes and waiting for a human operator to bring them all back up.

If we don't automate this inside Scylla, we can automate it outside, i.e., https://github.com/scylladb/scylladb/issues/8631 - the external tool I keep dreaming about could tell a human operator in 5 minutes: "Node X's I/O isn't functioning, please consider rebooting or replacing node X quickly. We could then have restored service to this user in several minutes instead of several hours.

> Even worse is when you have a large slowdown. Then the detector thinks everything is okay but in reality the entire cluster is slowed down to the speed of the slowest disk or node.

I agree. Again this situation can be automatically detected using metrics, so perhaps again an external tool should detect and report it instead of relying on a rusty developer like me who only debugs such problems once in a blue moon will come and think that this might be the issue.
 
> I don't think we'll be able to make progress here.

Perhaps.  I can close this issue and raise the specific checker's idea in https://github.com/scylladb/scylladb/issues/8631 (and perhaps start thinking about actually doing #8631 instead of just talking about it).

**Comment by avikivity** [comment]
Saying something immediately is of course possible, but it's easy to find a transient reason for the problem.
